subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -674,36 +674,32 @@ private ResolveState resolveToStateOrLater(final InternalState requestedState) {
                         .willBecomeAnErrorInGradle9()
                         .withUserManual("viewing_debugging_dependencies", "sub:resolving-unsafe-configuration-resolution-errors")
                         .nagUser();
-                return domainObjectContext.getModel().fromMutableState(p -> resolveExclusively(requestedState));
+                newState = domainObjectContext.getModel().fromMutableState(p -> resolveExclusivelyIfRequired());
             }
+        } else {
+            newState = resolveExclusivelyIfRequired();
         }
-        return resolveExclusively(requestedState);
+
+        return newState.getCachedResolverResults();
     }
 
-    private ResolveState resolveExclusively(InternalState requestedState) {
-        return currentResolveState.update(initial -> {
-            ResolveState current = initial;
-            if (requestedState == GRAPH_RESOLVED || requestedState == ARTIFACTS_RESOLVED) {
-                current = resolveGraphIfRequired(requestedState, current);
-            }
-            if (requestedState == ARTIFACTS_RESOLVED) {
-                current = resolveArtifactsIfRequired(current);
+    private ResolveState resolveExclusivelyIfRequired() {
+        return currentResolveState.update(currentState -> {
+            if (currentState.getState() == GRAPH_RESOLVED) {
+                if (dependenciesModified) {
+                    throw new InvalidUserDataException(String.format("Attempted to resolve %s that has been resolved previously.", getDisplayName()));
+                }
+                return currentState;
             }
-            return current;
+
+            return resolveGraphInBuildOperation();
         });
     }
 
     /**
-     * Must be called from {@link #resolveExclusively(InternalState)} only.
+     * Must be called from {@link #resolveExclusivelyIfRequired} only.
```

Comment:
🤔 There's only 1 call site?  Why not just inline it there and not worry about people misusing it from elsewhere?

Reply:
While we could do this I would rather not, as this is a very long method as-is and the method that calls this is already 3 indents deep. I try not to have such deeply-nested code. 

I agree though this is a little weird, but since it was already like this I think its best to keep for now. 

I will see if I can revisit this in future passes to see if it can be cleaned up